chanfitness/chanevent.go

```diff
@@ -0,0 +1,179 @@
+		return "peer_offline"
+	}
+
+	return "unknown"
+}
+
+// channelEvent is a a timestamped event which is observed on a per channel basis.
+type channelEvent struct {
+	timestamp time.Time
+	eventType eventType
+}
+
+// chanEventLog stores all events that have occurred over a channel's lifetime.
+type chanEventLog struct {
+	// id is the uint64 of the short channel ID.
+	id uint64
+
+	// peer is the compressed public key of the peer being monitored.
+	peer [33]byte
+
+	// events is a log of timestamped events observed for the channel.
+	events []*channelEvent
+
+	// now is expected to return the current time. It is supplied as an
+	// external function to enable deterministic unit tests.
+	now func() time.Time
+
+	// closed indicates whether a channel is closed, and is used to determine
+	// whether we should continue to track peer events for the channel without
+	// having to iterate through all events. When channel scored are persisted,
+	// closed channels can be removed from memory rather than having a marker.
+	closed bool
+}
+
+// add appends an event with the given type and current time to the event log.
+func (e *chanEventLog) add(eventType eventType) {
+	e.events = append(e.events, &channelEvent{
+		timestamp: e.now(),
+		eventType: eventType,
+	})
+
+	if eventType == channelClosedEvent {
+		e.closed = true
+	}
+
+	log.Debugf("Channel %v recording event: %v", e.id, eventType)
+}
+
+// getLifespan returns the total observed lifetime of a channel's event log.
+// If a channel is open, the lifespan is calculated from open time until the
+// present. If the channel is closed, this lifespan terminates when the channel
+// was closed. In absence of a channel opened event, the earliest entry in the
+// event log is considered to be the opening time.
+func (e *chanEventLog) getLifespan() time.Duration {
+	// If there are no events, we cannot calculate lifetime.
+	if len(e.events) == 0 {
+		return 0
+	}
+
+	// Close time starts as the present time, but will be set back to the actual
+	// closing time if there is a close event present.
+	closeTime := e.now()
+
+	var openTime time.Time
+
+	for _, e := range e.events {
+		// Set open time to the earliest timestamp in the event log. The earliest
+		// event is expected to be a channel opened event, but if it is not, it
+		// will be overridden by a subsequent open event, should one be present,
+		// because the events log is ordered by time for this loop.
+		if openTime.IsZero() || e.timestamp.Before(openTime) {
+			openTime = e.timestamp
+		}
+
+		switch e.eventType {
+		case channelOpenedEvent:
+			openTime = e.timestamp
+
+		case channelClosedEvent:
+			closeTime = e.timestamp
+		}
+
+	}
+
+	// If a close event was recorded before an open event, just return zero to
+	// prevent returning a negative lifespan.
+	if closeTime.Before(openTime) {
+		return 0
+	}
+
+	return closeTime.Sub(openTime)
+}
+
+// getUptime returns total uptime of a channel. Online periods are calculated
```

Comment:
Question from the user perspective: how would a user interpret the up time of an relatively old channel? Suppose it is 5 months old. The first 4 months the up time was 100%, the last month it was offline all the time. This would be reported as 80%, which may sound better than what the fitness of the channel really is.

I think the scope of the pr as it is is fine, but it might be good to give this a thought nevertheless.

Reply:
That could be addressed by adding a `startTime`/`endTime` param to the uptime calculation? 
For total uptime you just pass the open time/now and for more specific queries you can provide time spans within the channel's lifetime.

The user would have to get lifetime, then query specific uptime 'buckets' which is a bit roundabout, but a specialised endpoint which provides uptime per some duration chunk could be added? Eg give me uptime % per hour/day.

Reply:
Yes, `startTime`/`endTime` sounds good.

Maybe eventually a single call can develop that returns the channel fitness within a given time frame. One of the fitness parameters that is returned, is the channel open duration within the time frame, so that the percentage calculation can still be made correctly.

And then indeed `listChannels` could show three sets of fitness values for let's say 'last day', 'last week' and 'all'.